pds4indextools/pds4_create_xml_index.py

```diff
@@ -198,271 +701,781 @@ def update_nillable_elements_from_xsd_file(xsd_file, nillable_elements_info):
+        # associated prefixes.
         namespaces = root.nsmap
         namespaces['pds'] = namespaces.pop(None)
         prefixes = {v: k for k, v in namespaces.items()}
 
-        xml_results = {}
-        traverse_and_store(root, tree, xml_results, elements_to_scrape,
-                           nillable_elements_info, config, file)
-
-        for key in list(xml_results.keys()):
-            process_tags(xml_results, key, root,
-                         namespaces, prefixes, args)
+        # Each label file is now traversed and scraped for the requested content.
+        label_results = {}
+        traverse_and_store(root, tree, label_results, xsd_files,
+                           nillable_elements_info, config, label_file)
+
+        # The XPath headers in the label_results dictionary are reformatted to
+        # improve readability. Each XPath's namespace is replaced with its prefix for
+        # faster reference. Duplicate XPaths are made unique to ensure all results are
+        # present in the final product.
+        for key in list(label_results.keys()):
+            process_headers(label_results, key, root, namespaces, prefixes)
+
+        for key in list(label_results.keys()):
+            key_new = key.replace('[', '<')
+            key_new = key_new.replace(']', '>')
+            label_results[key_new] = label_results.pop(key)
+
+        for key in list(label_results.keys()):
+            parts = key.split('/')
+            new_parts = []
+            for part in parts:
+                if not part.endswith('>') and parts.index(part) != 1:
+                    part = part+'<1>'
+                    new_parts.append(part)
+                else:
+                    new_parts.append(part)
+            key_new = '/'.join(new_parts[1:])
+            label_results[key_new] = label_results.pop(key)
+
+        for key in list(label_results.keys()):
+            if 'cyfunction' in key:
+                del label_results[key]
+
+        # The XPath headers must be renumbered to reflect which instance of the element
+        # the column refers to. At this stage, duplicate XPaths may exist again due to
+        # the reformatting. These duplicates are corrected to preserve the contents of
+        # each element's value.
+        xpath_map = renumber_xpaths(label_results.keys())
+        for old_xpath, new_xpath in xpath_map.items():
+            label_results[new_xpath] = label_results.pop(old_xpath)
+
+        correct_duplicates(label_results)
+
+        # The label_results dictionary will now be filtered according to the contents
+        # of the --limit-xpaths-file input file. If this command is not used, the original
+        # dictionary will be returned. Glob patterns are processed sequentially, with the
+        # first pattern having the highest priority.
+        label_results = filter_dict_by_glob_patterns(
+            label_results, elements_to_scrape, verboseprint)
+
+        if label_results == []:
```

Comment:
I'm always more of a fan of `len(label_results) == 0`. The reason is that if you say `[]` explicitly, if in the future you changed `filter_dict_by_glob_patterns` to return a tuple or something else, which wouldn't otherwise break the code, it *would* break this test since it isn't a list anymore. Of course the pure Python way to do it is `if not label_results:`

Reply:
For some reason, a single unit test is failing with this change. And I really can't tell why. `label_results` will have content in it, but it triggers the 'No results found: glob pattern(s) excluded all matches' message.

Reply:
That's a case where you should dig deeper to see what the problem is. Unexplained behavior can point to a hidden bug. Don't just go back to the code that "works" if you don't know why the other doesn't work. Try printing out the content of label_results just before this test and then seeing what the output is when running the tests.